Modules/itertoolsmodule.c

```diff
@@ -3802,6 +3802,75 @@ static PyType_Spec ziplongest_spec = {
 };
 
 
+/* ilen function *************************************************************/
+
+/*[clinic input]
+itertools.ilen
+
+    iterable: object
+    /
+
+Equivalent to len(list(iterable)).
+[clinic start generated code]*/
+
+static PyObject *
+itertools_ilen(PyObject *module, PyObject *iterable)
+/*[clinic end generated code: output=99a1c852a6d09dd4 input=cf51dec933b693da]*/
+{
+    PyObject *it, *item;
+    PyObject *(*iternext)(PyObject *);
+    PyObject *long_cnt = NULL;
+    PyObject *new_long_cnt, *temp;
+    Py_ssize_t cnt;
+
+    it = PyObject_GetIter(iterable);
```

Comment:
I'm just wondering, but is it possible to know whether an object is an iterable with known size or not in advance? for instance, if you pass `ilen([1,2,3])`, it should be equivalent to directly call `len` instead of having to count (again) the elements, right?

Reply:
It is, but I don't think it is a good idea. It complicates code unnecessarily.

It is iterator module. If one has a list, he should call `len`.

E.g. `islice` function could jump straight to the index if a list is provided, but it doesn't. Because it is not what it is supposed to do.

It is good to be able to count on it that it will always call `iter` and iterate.

Reply:
So maybe it should be good to say so? like add a recommendation about it (in the docs)

Reply:
I have adjusted documentation to be more clear as per your suggestion above. Will push in a minute, let me know what you think. If it is ok or explicit recommendation to use `len` for sequences needed.